File: typescript/systemdesign/delayedJobScheduler.ts
/*
Delayed job scheduler: jobs are enqueued with an execution timestamp and
run in that order via a hand-rolled binary min-heap keyed by execution
time, so the next-due job is always found/removed in O(log n).

status - completed
*/

import { logp } from "../utils/logger";

interface ScheduledJob {
  id: number;
  runAtMs: number;
  name: string;
}

class MinHeap<T> {
  private readonly items: T[] = [];

  constructor(private readonly compare: (a: T, b: T) => number) {}

  get size(): number {
    return this.items.length;
  }

  peek(): T | undefined {
    return this.items[0];
  }

  push(item: T): void {
    this.items.push(item);
    this.bubbleUp(this.items.length - 1);
  }

  pop(): T | undefined {
    if (this.items.length === 0) {
      return undefined;
    }
    const top = this.items[0];
    const last = this.items.pop()!;
    if (this.items.length > 0) {
      this.items[0] = last;
      this.bubbleDown(0);
    }
    return top;
  }

  private bubbleUp(index: number): void {
    let i = index;
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (this.compare(this.items[i], this.items[parent]) >= 0) {
        break;
      }
      [this.items[i], this.items[parent]] = [this.items[parent], this.items[i]];
      i = parent;
    }
  }

  private bubbleDown(index: number): void {
    let i = index;
    const n = this.items.length;
    while (true) {
      const left = 2 * i + 1;
      const right = 2 * i + 2;
      let smallest = i;
      if (left < n && this.compare(this.items[left], this.items[smallest]) < 0) {
        smallest = left;
      }
      if (right < n && this.compare(this.items[right], this.items[smallest]) < 0) {
        smallest = right;
      }
      if (smallest === i) {
        break;
      }
      [this.items[i], this.items[smallest]] = [this.items[smallest], this.items[i]];
      i = smallest;
    }
  }
}

class DelayedJobScheduler {
  private readonly heap = new MinHeap<ScheduledJob>((a, b) => a.runAtMs - b.runAtMs);
  private nextId = 1;

  schedule(name: string, runAtMs: number): number {
    const id = this.nextId++;
    this.heap.push({ id, runAtMs, name });
    return id;
  }

  // runs every job whose runAtMs has arrived by nowMs, earliest first
  runDueJobs(nowMs: number): ScheduledJob[] {
    const executed: ScheduledJob[] = [];
    while (this.heap.size > 0 && this.heap.peek()!.runAtMs <= nowMs) {
      executed.push(this.heap.pop()!);
    }
    return executed;
  }

  pendingCount(): number {
    return this.heap.size;
  }
}

// usage scenarios

const scheduler = new DelayedJobScheduler();
scheduler.schedule("send-welcome-email", 300);
scheduler.schedule("expire-session", 100);
scheduler.schedule("send-reminder", 200);
scheduler.schedule("cleanup-temp-files", 500);

logp(`pending jobs before any run: ${scheduler.pendingCount()}`);

const dueAt250 = scheduler.runDueJobs(250);
logp(`jobs due by t=250: ${dueAt250.map((j) => `${j.name}@${j.runAtMs}`).join(", ")}`);
logp("expect expire-session(100) then send-reminder(200), in that time order");

logp(`pending jobs remaining: ${scheduler.pendingCount()}`);

const dueAt1000 = scheduler.runDueJobs(1000);
logp(`jobs due by t=1000: ${dueAt1000.map((j) => `${j.name}@${j.runAtMs}`).join(", ")}`);
logp(`pending jobs remaining: ${scheduler.pendingCount()}`);

File: python/systemdesign/delayed_job_scheduler.py
"""
Delayed job scheduler: jobs are scheduled to run after a delay and
kept in a min-heap ordered by their execution time, so the scheduler
can cheaply peek and pop the next job that's due without scanning the
whole queue.

status - completed
"""

import heapq
import itertools
import time

from base_logger.logging_event import create_logger


class DelayedJobScheduler:

    log = create_logger(__name__)

    def __init__(self, clock=time.monotonic):
        self.clock = clock
        self.heap = []
        self._counter = itertools.count()

    def schedule(self, job, delay_seconds):
        run_at = self.clock() + delay_seconds
        entry_id = next(self._counter)
        heapq.heappush(self.heap, (run_at, entry_id, job))
        return entry_id

    def next_run_at(self):
        return self.heap[0][0] if self.heap else None

    def run_due_jobs(self):
        now = self.clock()
        results = []
        while self.heap and self.heap[0][0] <= now:
            run_at, entry_id, job = heapq.heappop(self.heap)
            results.append(job())
        return results

    def pending_count(self):
        return len(self.heap)


if __name__ == "__main__":
    log = create_logger("delayed_job_scheduler_demo")

    fake_time = [0.0]

    def fake_clock():
        return fake_time[0]

    scheduler = DelayedJobScheduler(clock=fake_clock)

    scheduler.schedule(lambda: "job-A (delay 10s)", delay_seconds=10)
    scheduler.schedule(lambda: "job-B (delay 3s)", delay_seconds=3)
    scheduler.schedule(lambda: "job-C (delay 7s)", delay_seconds=7)

    log.info("3 jobs scheduled, pending=%d, next_run_at=%s",
             scheduler.pending_count(), scheduler.next_run_at())

    fake_time[0] = 4.0
    ran = scheduler.run_due_jobs()
    log.info("at t=4s, jobs run: %s (expected only job-B, delay 3s)", ran)

    fake_time[0] = 8.0
    ran = scheduler.run_due_jobs()
    log.info("at t=8s, jobs run: %s (expected only job-C, delay 7s)", ran)

    fake_time[0] = 11.0
    ran = scheduler.run_due_jobs()
    log.info("at t=11s, jobs run: %s (expected job-A, delay 10s)", ran)

    log.info("pending jobs remaining: %d (expected 0)", scheduler.pending_count())
